cli/internal/lockfile/pnpm_lockfile.go

```diff
@@ -0,0 +1,321 @@
+func (p *PnpmLockfile) ResolvePackage(name string, version string) (string, string, bool) {
+	resolvedVersion, ok := p.resolveSpecifier(name, version)
+	if !ok {
+		return "", "", false
+	}
+	key := fmt.Sprintf("/%s/%s", name, resolvedVersion)
+	if entry, ok := (p.Packages)[key]; ok {
+		var version string
+		if entry.Version != "" {
+			version = entry.Version
+		} else {
+			version = resolvedVersion
+		}
+		return key, version, true
+	}
+
+	return "", "", false
+}
+
+// AllDependencies Given a lockfile key return all (dev/optional/peer) dependencies of that package
+func (p *PnpmLockfile) AllDependencies(key string) (map[string]string, bool) {
+	deps := map[string]string{}
+	entry, ok := (p.Packages)[key]
+	if !ok {
+		return deps, false
+	}
+
+	for name, version := range entry.Dependencies {
+		deps[name] = version
+	}
+
+	for name, version := range entry.OptionalDependencies {
+		deps[name] = version
+	}
+
+	for name, version := range entry.PeerDependencies {
+		deps[name] = version
+	}
+
+	return deps, true
+}
+
+// Subgraph Given a list of lockfile keys returns a Lockfile based off the original one that only contains the packages given
+func (p *PnpmLockfile) Subgraph(packages []string) (Lockfile, error) {
+	lockfilePackages := make(map[string]PackageSnapshot, len(packages))
+	for _, key := range packages {
+		entry, ok := p.Packages[key]
+		if ok {
+			lockfilePackages[key] = entry
+		} else {
+			return nil, fmt.Errorf("Unable to find lockfile entry for %s", key)
+		}
+	}
+
+	lockfile := PnpmLockfile{
+		Version:            p.Version,
+		Importers:          p.Importers,
+		Packages:           lockfilePackages,
+		NeverBuiltDeps:     p.NeverBuiltDeps,
+		OnlyBuiltDeps:      p.OnlyBuiltDeps,
+		Overrides:          p.Overrides,
+		PackageExtChecksum: p.PackageExtChecksum,
+		PatchedDeps:        p.PatchedDeps,
+	}
+
+	return &lockfile, nil
+}
+
+// Encode encode the lockfile representation and write it to the given writer
+func (p *PnpmLockfile) Encode(w io.Writer) error {
+	if err := isSupportedVersion(p.Version); err != nil {
+		return err
+	}
+
+	funcMap := template.FuncMap{
+		"displayProjectSnapshot": displayProjectSnapshot,
+		"displayPackageSnapshot": displayPackageSnapshot,
+	}
+	t, err := template.New("pnpm lockfile").Funcs(funcMap).Parse(pnpmLockfileTemplate)
+	if err != nil {
+		return errors.Wrap(err, "unable to parse pnpm-lock.yaml template")
+	}
+
+	if p.hasCRLF {
+		var b bytes.Buffer
+		if err := t.Execute(&b, p); err != nil {
+			return err
+		}
+		scanner := bufio.NewScanner(&b)
+		crlf := []byte("\r\n")
+		for scanner.Scan() {
+			line := scanner.Bytes()
+			if _, err := w.Write(line); err != nil {
+				return errors.Wrap(err, "error copying lockfile line")
+			}
+			if _, err := w.Write(crlf); err != nil {
+				return errors.Wrap(err, "error copying lockfile line")
+			}
+		}
+		if err := scanner.Err(); err != nil {
+			return errors.Wrap(err, "error reading lockfile line from buffer")
+		}
+
+		return nil
+	}
```

Comment:
Yowch! I really, really hope this doesn't have to stick around. /me yells at Windows.

Reply:
I really dislike this as well. Due to the template usage this ended up being the least 🤮 of the solutions I came up with.

Reply:
I quite like to use https://github.com/andybalholm/crlf as a wrapper for the CRLF issues :)

Reply:
Thanks for that suggestion, that's so much nicer!